## Задача 3-2. Задача TSP: нижняя оценка Гельда—Карпа.

В этой задаче Вам предлагается релизовать алгоритм Гельда—Карпа для нижней оценки стоимости решения в задаче Euclidean TSP.

Сделайте следующее:
* Скачайте файл [`tsp-instances.zip`](https://github.com/dainiak/discrete-optimization-course/raw/master/tsp-instances.zip) и разархивируйте из него файлы со входами задачи TSP. Это в точности те же входные данные, что и в задании 3-1.
* Реализуйте функцию `lower_bound_tsp`. При этом можно пользоваться каким-нибудь стандартным алгоритмом построения минимального остовного дерева из библиотеки [`networkx`](https://networkx.github.io/), входящей в состав дистрибутива Anaconda.
* Запустите функцию `run_all()`, чтобы протестировать свой код, и напишите полученные, как следствия, верхние оценки погрешностей решений, которые были получены Вашими алгоритмами NN и NI при решении задания 3-1. Запишите свои выводы в 1-2 предложениях в последней ячейке ipynb-файла.

In [4]:
from typing import List, Tuple
from math import sqrt
from itertools import combinations, islice
from networkx import  minimum_spanning_tree, parse_edgelist, minimum_spanning_edges
import numpy as np

def read_tsp_instance(filename: str) -> List[Tuple[int,int]]:
    with open(filename, 'r') as file:
        coordinates = []
        for line in file:
            line = line.strip().lower()
            if line.startswith('dimension'):
                coordinates = [(0, 0)] * int(line.split()[-1])
            tokens = line.split()
            if len(tokens) == 3 and tokens[0].isdecimal():
                tokens = line.split()
                coordinates[int(tokens[0])-1] = tuple(map(float, tokens[1:]))
        return coordinates


def euclidean_distance(point1: Tuple[int,int], point2: Tuple[int,int]) -> float:
    return sqrt((point1[0]-point2[0]) ** 2 + (point1[1]-point2[1]) ** 2)

In [122]:
def lower_bound_tsp(count, vertex_coordinates: List[Tuple[int,int]]) -> float:
    
    def get_graph():
        lines = []
        for i in range(1, len(vertex_coordinates) + 1): # Переводим заданный граф в формат, нужный стандартным функциям
            for j in range(1, len(vertex_coordinates) + 1):
                lines.append( str(i) + " " + str(j) + " {'weight':" + 
                         str(euclidean_distance(vertex_coordinates[i - 1], vertex_coordinates[j - 1])) + '}')
        Gr = parse_edgelist(lines, nodetype = int) # Граф, к которому уже применимы стандартные функции.
        return Gr
    
    
    def get_tree_weight(G):
        T = minimum_spanning_edges(G)
        tree = list(T)
        weight = 0
        for edge in tree: # Вычисляем вес минимального остовного дерева, складывая веса всех его рёбер.
            weight += euclidean_distance(vertex_coordinates[edge[0] - 1], vertex_coordinates[edge[1] - 1])
        return tree, weight
    

    l = []
    size = len(vertex_coordinates)
    G = get_graph() # Считываем граф
    G1 = G.copy()
    T, best_weight = get_tree_weight(G) # Вес остовного дерева.
    y = np.zeros(size) # "Потенциалы".
    two = np.ones_like(2)
    alpha = 1
    counts = np.zeros(size) # Степени вершин в дереве.
    weight = best_weight
    U = 2 * best_weight # Верхняя оценка. Как известно, длина оптимума не превосходит удвоенного веса остовного дерева.
    t = 100
    for times in range(0, count): # Количество шагов алгоритма.

        counts = np.zeros(size)
        for fr in T : # Подсчитываем степени вершин дерева.
            counts[fr[0] - 1] += 1
            counts[fr[1] - 1] += 1
        t = alpha *(U - weight)/(( two - counts)*( two - counts)).sum() # Пересчитываем "шаг" алгоритма.
        alpha *= 2/3 # Уменьшаем параметр каким-то произвольным образом.
        
        for i in range(0, size): # Пересчитываем потенциалы.
            y[i] += t * (2 - counts[i])
        
        for fr in range(1, size + 1): # Обновляем веса рёбер.
            for to in range(1, size + 1):
                if G.has_edge(fr, to):
                    G1[fr][to]['weight'] = G[fr][to]['weight'] - y[fr-1]
        
        T, weight = get_tree_weight(G1)# Оценка.
        weight += 2 * y.sum() 
        if(weight > best_weight):
            best_weight = weight
   
    return(best_weight)

In [123]:
import time
from os.path import exists

def run_all():
    # Порядок следования графов изменён!
    instance_filenames = [['d198.tsp', 50], ['d493.tsp', 50], ['d657.tsp', 15], 
                          ['d2103.tsp', 3], ['pr107.tsp', 15], ['pr152.tsp',15],
                          ['pr439.tsp', 15]]
    for filename, count in instance_filenames:
        if not exists(filename):
            print('File not found: “{}”. Skipping this instance.'.format(filename))
            continue
        instance = read_tsp_instance(filename)
        print('Instance {}…'.format(filename), end='')
        time_start = time.monotonic()
        bound = lower_bound_tsp(count, instance)
        time_nn = time.monotonic()-time_start
        print(' done in {:.2} seconds with lower bound {}'.format(time_nn, int(bound)))

In [124]:
run_all()

Instance d198.tsp… done in 2.5e+01 seconds with lower bound 14007
Instance d493.tsp… done in 9.8e+01 seconds with lower bound 33384
Instance d657.tsp… done in 6.9e+01 seconds with lower bound 49379
Instance d2103.tsp… done in 4e+02 seconds with lower bound 78979
Instance pr107.tsp… done in 1.5 seconds with lower bound 42015
Instance pr152.tsp… done in 2.9 seconds with lower bound 68086
Instance pr439.tsp… done in 3.3e+01 seconds with lower bound 106352


## Выводы
Запишите здесь полученные результаты относительно погрешностей алгоритмов NN и NI.

Из предыдущего задания 3-1 имеем маршруты:

1

Solving instance d198.tsp... with tour length 18506 using NN and ... with tour length 19944 using NI

Instance d198.tsp... done ... with lower bound 14007

Найденные значения на около 32% и 42%  для NN и NI соответственно больше оценки


2 

Solving instance d493.tsp... with tour length 43699 using NN and ... s with tour length 48762 using NI

Instance d493.tsp ... done ... with lower bound 33384

Найденные значения на около 31% и  46% для NN и NI соответственно больше оценки

3

Solving instance d657.tsp... with tour length 65173 using NN and ... with tour length 73468 using NI

Instance d657.tsp… done ... with lower bound 49379

Найденные значения на около 32% и 48%  для NN и NI соответственно больше оценки

4

Solving instance d2103.tsp… done...with tour length 94185 using NN and ...with tour length 137254 using NI

Instance d2103.tsp… done ... seconds with lower bound 78979

Найденные значения на около 19% и 74%  для NN и NI соответственно больше оценки 

5

Solving instance pr107.tsp… done... with tour length 51452 using NN and...with tour length 54492 using NI

Instance pr107.tsp… done... with lower bound 42015

Найденные значения на около 22% и 30%  для NN и NI соответственно больше оценки

6

Solving instance pr152.tsp… done... with tour length 88823 using NN and ...with tour length 106938 using NI

Instance pr152.tsp… done ... with lower bound 68086

Найденные значения на около 30% и 57%  для NN и NI соответственно больше оценки

7

Solving instance pr439.tsp… done... with tour length 131294 using NN and ... with tour length 173907 using NI

Instance pr439.tsp… done... with lower bound 106352

Найденные значения на около 23% и 64%  для NN и NI соответственно больше оценки


**Вывод:** на данных примерах жадные алгоритмы ошибаются в среднем на 30-50%, для этих примеров ошибка никогда не превышала более чем вдвое нижнюю оценку. 